simulide: init at 0.4.15-SR10

CarlosCraveiro opened the pull request:
## Description of changes
closes #273621 

Adds the version 0.4.15-SR10 of SimulIDE Circuit Simulator, a simple real time electronic circuit simulator, intended for hobbyist or students to learn and experiment with analog and digital electronic circuits and microcontrollers.
It supports PIC, AVR , Arduino and other MCUs and MPUs.

homepage URL: https://simulide.com/

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

loco-choco commented on pkgs/applications/science/electronics/simulide/default.nix:
```diff
@@ -0,0 +1,55 @@
+
+in
+  symlinkJoin {
+    name = "${pname}_${version}-${release}";
```
While trying the package with `nix run`, it tries to run it with the name "simulide_0.4.15-SR10", but the name of the binary is "simulide", which results in an error. The issue is probably this name in symlinkJoin not matching the binary.

CarlosCraveiro replied:
Hmm, sorry about that. I forgot to execute ```nix run```, so I haven't seen this problem earlier.

loco-choco commented on pkgs/applications/science/electronics/simulide/default.nix:
```diff
@@ -0,0 +1,55 @@
+      cp ${desktopItem}/share/applications/* $out/share/applications/
+    '';
+
+    meta = with lib; {
```
Might be good to add the `mainProgram` on meta.

CarlosCraveiro replied:
Thanks for the suggestion! I applied the changes and now it's working!!

loco-choco commented:
LGTM!

WolfangAukang commented on pkgs/applications/science/electronics/simulide/default.nix:
```diff
@@ -0,0 +1,56 @@
+       '';
+      homepage = "https://simulide.com/";
+      mainProgram = "simulide";
+      license = licenses.gpl3;
```
This license does not exist. Please check [here](https://github.com/NixOS/nixpkgs/blob/master/lib/licenses.nix).

CarlosCraveiro replied:
Sorry about that, but now I corrected it. It is ```licences.gpl3Only```.

WolfangAukang commented on pkgs/applications/science/electronics/simulide/default.nix:
```diff
@@ -0,0 +1,56 @@
+
+  src = fetchurl {
+    url = "https://launchpad.net/${pname}/${version}/${version}-stable/+download/SimulIDE_${version}-${release}.AppImage";
+    sha256 = "25f85f5262b39c8b25da042f11a0fd517c5d2750e81e538ac15992659ebea03c";
```
```suggestion
    hash = "sha256-JfhfUmKznIsl2gQvEaD9UXxdJ1DoHlOKwVmSZZ6+oDw=";
```
Please use the hash in SRI format by doing the following:
```
> nix hash to-sri --type sha256 25f85f5262b39c8b25da042f11a0fd517c5d2750e81e538ac15992659ebea03c
sha256-JfhfUmKznIsl2gQvEaD9UXxdJ1DoHlOKwVmSZZ6+oDw=
```

CarlosCraveiro replied:
Ok, I just changed the hash.

WolfangAukang commented on pkgs/top-level/all-packages.nix:
```diff
@@ -39843,6 +39843,8 @@ with pkgs;
 
   qucs-s = qt6Packages.callPackage ../applications/science/electronics/qucs-s { };
 
+  simulide = callPackage ../applications/science/electronics/simulide { };
```
For new packages, you need to add them to `pkgs/by-name` instead. Please check [here](https://github.com/NixOS/nixpkgs/blob/master/pkgs/by-name/README.md). It won't be necessary to add the new package to this file.

CarlosCraveiro replied:
I added my package to pkgs/by-name. It seems it is building fine by doing:

```bash
nix-build -A simulide
```

But i don't know why now the new syntax ``` nix run .#simulide``` and ```nix build #simulide``` stoped working. Is it some limitation of ```pkgs/by-name```?

WolfangAukang reviewed (changes requested):
Thank you for contributing with nixpkgs. Some observations to consider.

WolfangAukang commented:
Also, did you try to package it through the source code? There are instructions on the [README](https://bazaar.launchpad.net/~arcachofo/simulide/trunk/view/head:/README.md) and it requires qt5.

CarlosCraveiro commented:
> Also, did you try to package it through the source code? There are instructions on the [README](https://bazaar.launchpad.net/~arcachofo/simulide/trunk/view/head:/README.md) and it requires qt5.

Actually I did not tried to package through the source code. I used the AppImage because it was the form I was using the program in a course I'm taking at college. But it is a good idea.

TomaSajt commented:
I think I have managed to build from source, but it needs a bit more work.
I will send the code here once I'm done, if you want to maintain it.

WolfangAukang commented:
@CarlosCraveiro can you please test @TomaSajt's file to see if it works for you? If it works, you can change your PR to use the indicated code instead. Some changes are needed, but as long it works and can be built from source, it is the recommended course of action.

CarlosCraveiro commented:
> Here's what I came up:
> ### `pkgs/by-name/si/simulide/package.nix`
> ```nix
> { lib
> , stdenv
> , fetchbzr
> , substituteAll
> , libsForQt5
> }:
> 
> stdenv.mkDerivation (finalAttrs: {
>   pname = "simulide";
>   version = "1.1.0-RC0";
> 
>   src = fetchbzr {
>     url = "https://code.launchpad.net/~arcachofo/simulide/trunk";
>     rev = "1910";
>     sha256 = "sha256-ZLB2y6ZMU67l3X9dXc+6LSyjLRdTslzESBGZhXYelxM=";
>   };
> 
>   patches = [
>     (substituteAll {
>       src = ./make-deterministic.patch;
>       rev_no = finalAttrs.src.rev;
>       build_date = "20-09-23"; # commit date of revision (dd-MM-yy)
>     })
>   ];
> 
>   preConfigure = ''
>     cd build_XX
>   '';
> 
>   nativeBuildInputs = [
>     libsForQt5.qmake
>     libsForQt5.wrapQtAppsHook
>   ];
> 
>   buildInputs = [
>     libsForQt5.qtserialport
>     libsForQt5.qtmultimedia
>     libsForQt5.qttools
>   ];
> 
>   installPhase = ''
>     runHook preInstall
> 
>     sed -i ../resources/simulide.desktop \
>       -e "s|^Exec=.*$|Exec=simulide|" \
>       -e "s|^Icon=.*$|Icon=simulide|"
>     install -Dm644 ../resources/simulide.desktop $out/share/applications/simulide.desktop
> 
>     install -Dm644 ../resources/icons/hicolor/256x256/simulide.png $out/share/icons/hicolor/256x256/apps/simulide.png
> 
>     mkdir -p $out/share/simulide $out/bin
>     cp -r executables/SimulIDE_*/* $out/share/simulide
>     mv $out/share/simulide/simulide $out/bin/simulide
> 
>     runHook postInstall
>   '';
> 
>   meta = {
>     description = "A simple real time electronic circuit simulator";
>     longDescription = ''
>       SimulIDE is a simple real time electronic circuit simulator, intended for hobbyist or students
>       to learn and experiment with analog and digital electronic circuits and microcontrollers.
>       It supports PIC, AVR, Arduino and other MCUs and MPUs.
>     '';
>     homepage = "https://simulide.com/";
>     license = lib.licenses.gpl3Only;
>     mainProgram = "simulide";
>     maintainers = with lib.maintainers; [ carloscraveiro tomasajt ];
>     platforms = lib.platforms.unix;
>   };
> })
> ```
> 
> ### pkgs/by-name/si/simulide/make-determinstic.nix
> ```patch
> diff --git a/SimulIDE.pro b/SimulIDE.pro
> index 2a6c881a..75248c89 100644
> --- a/SimulIDE.pro
> +++ b/SimulIDE.pro
> @@ -106,13 +106,13 @@ CONFIG += warn_on
>  CONFIG += no_qml_debug
>  CONFIG *= c++11
>  
> -REV_NO = $$system( bzr revno )
> +REV_NO = @rev_no@
>  DEFINES += REVNO=\\\"$$REV_NO\\\"
>  
>  DEFINES += MAINMODULE_EXPORT=
>  DEFINES += APP_VERSION=\\\"$$VERSION$$RELEASE\\\"
>  
> -BUILD_DATE = $$system($(which date) +\"\\\"%d-%m-%y\\\"\")
> +BUILD_DATE = @build_date@
>  DEFINES += BUILDDATE=\\\"$$BUILD_DATE\\\"
>  
>  TARGET_NAME   = SimulIDE_$$VERSION$$RELEASE
> ```
> 
> maybe the `make-deterministic` name should be changed to something like `set-date-and-revision`

Thanks a lot for the help!  I'll test the package today and update here!

CarlosCraveiro commented:
> @CarlosCraveiro can you please test @TomaSajt's file to see if it works for you? If it works, you can change your PR to use the indicated code instead. Some changes are needed, but as long it works and can be built from source, it is the recommended course of action.

Sure, I will test the package file as soon as possible and update here. Thanks a lot for the help and advices!

CarlosCraveiro commented:
@TomaSajt I tested the package you proposed, and everything seems working fine. However I want to package specifically the version 0.4.14 of simulide, because I is the latest version of simulide that comes with some versions of the PIC microcontroler like pic18f4550. I tried to change the source to https://code.launchpad.net/~arcachofo/simulide/simulide_0.4.14 , but I don't know how to fetch the code to generate the hash. Could you please help me with instructions about how do I do this?

TomaSajt commented:
Once you change anything about a fetcher, you'll need to also change the hash (only 1 character, or just set the hash as an empty string) to make nix realise that there's a new thing to download, it will say that you got a hash mismatch, but will say what the right one would be.

However the patch doesn't work on that older version, and there's a qtscript dependency also
I will take a look at this later today, to figure out how to handle the fact that people might not want to use the latest version.

TomaSajt commented:
Okay here's something that works with both versions:
```nix
{ lib
, stdenv
, fetchbzr
, substituteAll
, libsForQt5
, old ? false
}:

let
  versionTable = {
    old = {
      version = "0.4.15";
      release = "SR10";
      branch = "simulide_0.4.14"; # the branch name does not mach the version for some reason
      rev = "291";
      sha256 = "sha256-BBoZr/S2pif0Jft5wrem8y00dXl08jq3kFiIUtOr3LM=";
    };
    new = {
      version = "1.1.0";
      release = "RC0";
      branch = "1.1.0";
      rev = "1909";
      sha256 = "sha256-AvucduVraSlTXT8Hnpgh9i5Tsxji++vgjAJX1ZFnMs4=";
    };
  };
  versionInfo = if old then versionTable.old else versionTable.new;
in

stdenv.mkDerivation {
  pname = "simulide";
  version = "${versionInfo.version}-${versionInfo.release}";

  src = fetchbzr {
    url = "https://code.launchpad.net/~arcachofo/simulide/${versionInfo.branch}";
    inherit (versionInfo) rev sha256;
  };

  postPatch = ''
    sed -i resources/simulide.desktop \
      -e "s|^Exec=.*$|Exec=simulide|" \
      -e "s|^Icon=.*$|Icon=simulide|"
    sed -i SimulIDE.pro \
      -e "s|^VERSION = .*$|VERSION = ${versionInfo.version}|" \
      -e "s|^RELEASE = .*$|RELEASE = -${versionInfo.release}|" \
      -e "s|^REV_NO = .*$|REV_NO = ${versionInfo.rev}|" \
      -e "s|^BUILD_DATE = .*$|BUILD_DATE = ??-??-??|"
  '';

  preConfigure = ''
    cd build_XX
  '';

  nativeBuildInputs = [
    libsForQt5.qmake
    libsForQt5.wrapQtAppsHook
  ];

  buildInputs = [
    libsForQt5.qtserialport
    libsForQt5.qtmultimedia
    libsForQt5.qttools
  ] ++ lib.optional old libsForQt5.qtscript;

  installPhase = ''
    runHook preInstall

    install -Dm644 ../resources/simulide.desktop $out/share/applications/simulide.desktop
    install -Dm644 ../resources/icons/hicolor/256x256/simulide.png $out/share/icons/hicolor/256x256/apps/simulide.png

    mkdir -p $out/share/simulide $out/bin

    pushd executables/SimulIDE_*
    ${lib.optionalString old ''
      cp -r share/simulide/* $out/share/simulide
      cp bin/simulide $out/bin/simulide
    ''}
    ${lib.optionalString (!old) ''
      cp -r * $out/share/simulide
      mv $out/share/simulide/simulide $out/bin/simulide
    ''}
    popd

    runHook postInstall
  '';

  meta = {
    description = "A simple real time electronic circuit simulator";
    longDescription = ''
      SimulIDE is a simple real time electronic circuit simulator, intended for hobbyist or students
      to learn and experiment with analog and digital electronic circuits and microcontrollers.
      It supports PIC, AVR, Arduino and other MCUs and MPUs.
    '';
    homepage = "https://simulide.com/";
    license = lib.licenses.gpl3Only;
    mainProgram = "simulide";
    maintainers = with lib.maintainers; [ /* carloscraveiro */ tomasajt ];
    platforms = lib.platforms.unix;
  };
}
```
I don't know if there's a better way than this, if we want to stay in the `pkgs/by-name` structure. You'll need to override the `old` attribute to true for the older version.

We could do a thing other packages do: versioning in the package name. By this, I mean something like `simulide_0_4_15`, `simulide_1_0_0` and `simulide_1_1_0` are all in nixpkgs, with a generic builder, like what I have done here. And all we'd need to do is update the `versionTable` when there's a new release (like `-SR1` to `-SR2`)

I think the priority is to get it to state that can be committed, and then do some restructuring in another PR.

WolfangAukang commented:
@TomaSajt Another option (that could be developed in a separate PR) is using a generic structure to perform the update. [Here](https://i.ckie.dev/Zu4IQzD.txt) is a list of examples on Nixpkgs that could be used to develop it and make it easier to handle multiple versions with small changes

CarlosCraveiro commented:
@TomaSajt, I just finished testing the package and every thing works just fine! Also, I tested the package in my personal nix configuration and the flag is working properly!

About versioning the package by name, would it be something like this? (I took a look in the examples @WolfangAukang cited)
```bash
└── simulide
│       ├── generic.nix
│       ├── simulide_0_4_15.nix
│       ├── simulide_1_0_0.nix
│       └── simulide_1_1_0.nix
```
Where it would be using the all_packages.nix to point to the correct version (So doing the packaging by the old method)? Just so I know if I understood the idea. But, if it is, I agree that we should finish this PR first, and then open another one.

About any changes in the package code, I don't see any improvements to be made. Is there something to be fixed/improved for the PR to be approved?

TomaSajt commented:
> @TomaSajt, I just finished testing the package and every thing works just fine! Also, I tested the package in my personal nix configuration and the flag is working properly!
> 
> About versioning the package by name, would it be something like this? (I took a look in the examples @WolfangAukang cited)
> 
> ```shell
> └── simulide
> │       ├── generic.nix
> │       ├── simulide_0_4_15.nix
> │       ├── simulide_1_0_0.nix
> │       └── simulide_1_1_0.nix
> ```
> 
> Where it would be using the all_packages.nix to point to the correct version (So doing the packaging by the old method)? Just so I know if I understood the idea. But, if it is, I agree that we should finish this PR first, and then open another one.

I was thinking of something like this, yeah

TomaSajt reviewed (approved):
LGTM

TomaSajt reviewed (changes requested):
Oh, just realized, that the non `x86_64-linux` builds are failing.
Maybe `aarch64` is just not supported.
And `darwin` seems like it has some custom build instructions.
I'll look into it, but if we can't find a solution, `meta.platforms` should be set to only `x86_64-linux`

WolfangAukang commented on pkgs/by-name/si/simulide/package.nix:
```diff
@@ -0,0 +1,99 @@
+}:
+
+let
+  versionTable = {
+    old = {
+      version = "0.4.15";
+      release = "SR10";
+      branch = "simulide_0.4.14"; # the branch name does not mach the version for some reason
+      rev = "291";
+      sha256 = "sha256-BBoZr/S2pif0Jft5wrem8y00dXl08jq3kFiIUtOr3LM=";
+    };
+    new = {
+      version = "1.1.0";
+      release = "RC0";
+      branch = "1.1.0";
+      rev = "1909";
+      sha256 = "sha256-AvucduVraSlTXT8Hnpgh9i5Tsxji++vgjAJX1ZFnMs4=";
+    };
+  };
+  versionInfo = if old then versionTable.old else versionTable.new;
```
It would be preferrable to keep a single version. As said before, rather create a generic mechanism to handle multiple version later.

CarlosCraveiro commented:
@TomaSajt, as per @WolfangAukang's latest recommendation, would it be ok with you if we only package version 0.14.15-SR10 for now? Additionally, I configured the platforms to only support ```x86_64-linux```. I think in the next PR we can see how to support the other meta-platforms as well, but if it's not good for you, we could see how to make the custom build to support them.

TomaSajt reviewed (approved):
LGTM
I am fine with this version as well

TomaSajt commented:
By the way, I asked around on the forums, and it turns out their JIT compiler only works on x86_64.
I am still working on a way to fix the darwin build, but I am still having problem with forcing it to use the required `gcc` compiler instead of `clang`.

TomaSajt reviewed (changes requested):
Could you change the commit and PR message to have `0.4.15-SR10` as the version?

TomaSajt commented on pkgs/by-name/si/simulide/package.nix:
```diff
@@ -0,0 +1,80 @@
+{ lib
+, stdenv
+, fetchbzr
+, substituteAll
```
```suggestion
```
this attribute seems to be unused

CarlosCraveiro commented:
Happy New Year @TomaSajt ! How is it going? Sorry for not commenting on your last comments about the Darwin build. I added the version in the commit message and in the PR title and removed that unused attribute.

TomaSajt commented:
Happy new year! Thanks for the fixes
I have given up on the darwin support for now, but will ask around the community someday later if someone could help get gcc working properly on darwin.

WolfangAukang commented on pkgs/by-name/si/simulide/package.nix:
```diff
@@ -0,0 +1,79 @@
+    libsForQt5.wrapQtAppsHook
+  ];
+
+  buildInputs = [
+    libsForQt5.qtserialport
+    libsForQt5.qtmultimedia
+    libsForQt5.qttools
+    libsForQt5.qtscript
+  ];
```
```suggestion
  buildInputs = with libsForQt5; [
    qtserialport
    qtmultimedia
    qttools
    qtscript
  ];
```

TomaSajt replied:
After writing this, I agree now that this looks better

WolfangAukang commented on pkgs/by-name/si/simulide/package.nix:
```diff
@@ -0,0 +1,79 @@
+    cd build_XX
+  '';
+
+  nativeBuildInputs = [
+    libsForQt5.qmake
+    libsForQt5.wrapQtAppsHook
+  ];
```
```suggestion
  nativeBuildInputs = with libsForQt5; [
    qmake
    wrapQtAppsHook
  ];
```

WolfangAukang commented on pkgs/by-name/si/simulide/package.nix:
```diff
@@ -0,0 +1,79 @@
+  release = "SR10";
+  branch = "simulide_0.4.14"; # the branch name does not mach the version for some reason
+  rev = "291";
+  sha256 = "sha256-BBoZr/S2pif0Jft5wrem8y00dXl08jq3kFiIUtOr3LM=";
```
what if you use `hash` instead of `sha256`?

TomaSajt replied:
I believe there isn't a `hash` option for `fetchbzr`. (Maybe I could add it)
I searched around on git a bit more, and it looks like some people are downloading archives from launchpad. It looks like there are sources released [here](https://launchpad.net/simulide/0.4.15/0.4.15-stable) for *this* version.
Sadly, not all tags have the sources released in an archive, so I think that that `fetchbzr` will be more universal.

CarlosCraveiro replied:
As @TomaSajt  said apparently there isn't a `hash` option `fetchbzr`.  I tried to use `hash` when we migrate from appImage to compiling everything from source, but I ran into this problem.

WolfangAukang replied:
Ok, fair enough. You can mark it as resolved then.

WolfangAukang reviewed (commented):
Commit should be `simulide: init at 0.4.15-SR10`

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-already-reviewed/2617/1374

NickCao commented on pkgs/by-name/si/simulide/package.nix:
```diff
@@ -0,0 +1,79 @@
+    cd build_XX
+  '';
+
+  nativeBuildInputs = with libsForQt5; [
```
I think it's recommended to use `libsForQt5.callPackage` for qt5 packages, since it has a custom version of `mkDerivation` for use.

TomaSajt replied:
I was wondering whether QT based apps should try to use `pkgs/by-name` or stay with the usual `qt*.callPackage`.

I personally would go with the `pkgs/by-name` one, but I can see the benefit of having the qt deps also acting like any other package being imported.

IIRC the QT `mkDerivation` has `wrapQtAppsHook` `qtbase` automatically included, but maybe there's some other useful things about it that I don't remember

NickCao replied:
> I was wondering whether QT based apps should try to use `pkgs/by-name` or stay with the usual `qt*.callPackage`.

These two can co-exist, the package.nix stays in by-name, while an entry is added to all-packages.nix.

TomaSajt replied:
IIRC I tried that once but it gave me some CI check fails.
Maybe I just misremember.
I can't currently re-check all-packages, but I don't recall seeing direct imports from pkgs/by-name in there.

NickCao replied:
```
❯ rg by-name pkgs/top-level/all-packages.nix
824:  vcpkg-tool = callPackage ../by-name/vc/vcpkg-tool/package.nix {
2967:  doublecmd = callPackage ../by-name/do/doublecmd/package.nix {
6002:        ../by-name/me/meson/emulator-hook.sh
8995:  gtkclipblock = callPackage ../by-name/gt/gtkclipblock/package.nix { stdenv = gcc13Stdenv; };
<and many more>
```

There's a handful of them

NickCao replied:
Oh but none of them are using libsForQt5.callPackage?

NickCao replied:
> So just to see if I understood correctly... would it be better to migrate the packaging to the old way? and then make the call via libsForQt5.callPackage?

I think so.

CarlosCraveiro replied:
Ok, it ended up taking me a while, but it seems that calling the package in `pkgs/by-name/` directly from all-packages "works just fine". I even tested the application locally on my machine that pulls the `simulide` through an overlay and everything seems to "work as it should".

CarlosCraveiro replied:
The double quotes were due to the following error:
```bash
Run if result/bin/nixpkgs-check-by-name --base "$base" .; then
  if result/bin/nixpkgs-check-by-name --base "$base" .; then
    exit 0
  else
    exitCode=$?
    echo "To run locally: ./maintainers/scripts/check-by-name.sh $GITHUB_BASE_REF https://github.com/$GITHUB_REPOSITORY.git"
    exit "$exitCode"
  fi
  shell: /usr/bin/bash -e {0}
  env:
    mergedSha: a4e7e7803169340212267591c521158dfea0f190
    base: /tmp/tmp.5Ni4eCkKHd
    TMPDIR: /home/runner/work/_temp
pkgs.simulide: This attribute is manually defined (most likely in pkgs/top-level/all-packages.nix), which is only allowed if the definition is of the form `pkgs.callPackage pkgs/by-name/si/simulide/package.nix { ... }` with a non-empty second argument.
Validation failed, see above errors
To run locally: ./maintainers/scripts/check-by-name.sh master https://github.com/NixOS/nixpkgs.git
Error: Process completed with exit code 1.
```
Highlighting the following:
> This attribute is manually defined (most likely in pkgs/top-level/all-packages.nix), which is only allowed if the definition is of the form `pkgs.callPackage pkgs/by-name/si/simulide/package.nix { ... }` with a non-empty second argument.

 An error that is understandable, but that **causes me confusion** because wouldn't it be kind of the same as what I'm doing in practice? Because as far as I understand the use of `libsForQt5.callPackage` is precisely a way of passing arguments that `callPackage` by default would not pass.  Or am I mistaken?

CarlosCraveiro replied:
Anyway, I'm now moving the package to the old hierarchy, as this check is indeed failing. Just one moment...

CarlosCraveiro replied:
Okay, it passed the checks, and the package still seems to work as it should!

TomaSajt commented on pkgs/applications/science/electronics/simulide/default.nix:
```diff
@@ -0,0 +1,79 @@
+{ lib
+, stdenv
```
```suggestion
```

TomaSajt commented on pkgs/applications/science/electronics/simulide/default.nix:
```diff
@@ -0,0 +1,79 @@
+{ lib
+, stdenv
+, fetchbzr
+, libsForQt5
```
Now that the qt-type `callPackage` is used, the qt deps should be included like this.

```suggestion
, mkDerivation
, qmake
, qtserialport
, qtmultimedia
, qttools
, qtscript
```

TomaSajt commented on pkgs/applications/science/electronics/simulide/default.nix:
```diff
@@ -0,0 +1,79 @@
+  rev = "291";
+  sha256 = "sha256-BBoZr/S2pif0Jft5wrem8y00dXl08jq3kFiIUtOr3LM=";
+in
+stdenv.mkDerivation {
```
This will use the qt's custom `mkDerivation`
```suggestion
mkDerivation {
```

TomaSajt commented on pkgs/applications/science/electronics/simulide/default.nix:
```diff
@@ -0,0 +1,79 @@
+    cd build_XX
+  '';
+
+  nativeBuildInputs = with libsForQt5; [
+    qmake
+    wrapQtAppsHook
+  ];
```
```suggestion
  nativeBuildInputs = [
    qmake
  ];
```

TomaSajt reviewed (changes requested):
I requested some changes which should be done if `libsForQt5.callPackages` is used

CarlosCraveiro commented:
Okay, I just made the changes here. @TomaSajt

NickCao reviewed (approved):
diff LGTM, waiting for ofborg.

TomaSajt reviewed (changes requested):
.

TomaSajt commented on pkgs/applications/science/electronics/simulide/default.nix:
```diff
@@ -0,0 +1,82 @@
+    inherit rev sha256;
+  };
+
+  postPatch = ''
```
```suggestion
  postPatch = ''
    # GCC 13 needs this header explicitly included
    sed -i src/gpsim/value.h -e '1i #include <cstdint>'
    sed -i src/gpsim/modules/watchdog.h -e '1i #include <cstdint>'
```

TomaSajt replied:
I tried doing `env.CXXFLAGS = "-include cstdint";` but it didn't seem to work. So I came up with this instead, inspired by other packages that had to patch GCC 13

CarlosCraveiro replied:
Okay, change done! But just one question. How did you know it wasn't being effective?

TomaSajt replied:
I git rebased to a recent commit in master and tried to compile the program.

TomaSajt reviewed (changes requested):
There were some changes to the gcc13 default headers, so `cstdint` needs to be explicitly included.
You should probably do a rebase to a more recent commit (which has gcc13), so you can check if it works properly.

Commit:
maintainers: add CarlosCraveiro

Commit:
simulide: init at 0.4.15-SR10

CarlosCraveiro commented:
After a lot of suffering I got the rebase (I'll never open a PR on my fork's master again). Now it's time to fix the problem with gcc.

TomaSajt commented:
> After a lot of suffering I got the rebase (I'll never open a PR on my fork's master again). Now it's time to fix the problem with gcc.

Yeah, it's a pain working on your own fork's master 
I'd recommend never committing to it and just syncing it sometimes with upstream.
